support/docker/build.sh

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env bash
+
+#
+# This script is ran from outside the docker environment, to run itself into
+# the docker environment.
+#
+# It will automatically handle calling `nix-build` correctly, and unwrapping
+# the `result` symlink into a real directory.
+#
+# A `Tow-Boot--nix-store` volume will be automatically created. It will hold
+# the `/nix/store` as a local binary cache. Otherwise the toolchain must be
+# downloaded again every invocation.
+#
+# This is not a one-to-one replacement for Nix, but should be sufficient for
+# basic building locally without installing Nix globally, if Docker is deemed
+# more appropriate.
+#
+
+set -e
+set -u
+set -o pipefail
+PS4=" $ "
+
+_nix_docker() {
+	local STORE_VOLUME
+	STORE_VOLUME="$(docker volume create "Tow-Boot--nix-store")"
+
+	exec docker run \
+		--volume "$STORE_VOLUME:/nix" \
+		--volume "$PWD:/Tow-Boot" \
+		-it nixos/nix \
+		"$@"
+}
+
+if [ -e /Tow-Boot ]; then
+	REAL_UID="$1"; shift
+	REAL_GID="$1"; shift
+
+	cd /Tow-Boot
+
+	# Cleanup a leftover result (unclean and rude)
+	if [ -e result ]; then
+		rm -rf result
+	fi
+
+	# Call `nix-build`, showing the user the actual invocation.
+	(set -x
+	nix-build --cores 0 "$@"
+	)
+
+	# Unwrap the `result` symlink
+	out=$(readlink -f result)
```

Comment:
Capturing stdout of `nix-build` could be better than doing this, not reliant on only one derivation being built and `--no-out-link` or `-o` not being passed. I think it's also less racy?

Reply:
I'm concerned about multiple outputs though. And yes I hate that part.

Reply:
How about building outside the `/Tow-Boot` volume (so the directory where the result symlink ends up is owned purely by the container) and then copying everything that ends up in said directory to `/Tow-Boot` afterwards? This would further break passing paths (currently it behaves as one would expect if the script is invoked from the tow-boot checkout) but I'm questioning whether we should really pass the arguments to `nix-build` directly at all --- I feel like too much of the interface is broken by this script and allowing the user to pass only attrs from the top level of Tow-Boot might be better than passing everything through as is to `nix-build`.

Reply:
I wonder too. A goal I tried to achieve is that it should map as closely to Nix as possible, such that upgrading to using Nix proper just needs to change the invocation to `nix-build`.

Reply:
In that case, the script should probably respect the user's cwd. Passing a path will, in the current incarnation, only work as expected if the user _happens_ to be in the root of the Tow-Boot repo.

Reply:
*except* we're in the docker container, which means we can't know the actual CWD.

This could be done, though, by detecting the CWD relative to the root of the project, and only accept incantations inside the repo.

Reply:
Exactly, we're already wrapping docker in this shell script, so we can also work out where we are relative to the repo root and pass it using `docker run`'s `-w`.

Reply:
I think we have to sacrifice at least one part of the CLI for this to work without parsing the whole CLI.

We *can* build in another directory, but `./.` now refers to somewhere else, and thus `error: getting status of ...: No such file or directory`.

Given I prefer expanding all out links to allow building arbitrary `.nix` files, I think the concession will be to forcibly refer to `<Tow-Boot/default.nix>`. Anyone needing something more involved is free to edit the script, install Nix or use some other means. At least *everything else* on the CLI should be fine, other than referring to additional paths. Which is fine, since I dropped supporting paths for `--arg device` in #89.

> I'm using the angled brackets here to hide the implementation detail of the volume... Otherwise `/Tow-Boot/default.nix` would be used, which at this point basically has as much value.

Reply:
Fair enough. This was mostly about making sure the limitations are a conscious decision and not an unfortunate accident :)

>have our cake and eat it too

We could mount the entire host system into the container. But that would be quite rude.

Reply:
It wouldn't solve identifying new build outputs to unwrap! With the current strategy, the build needs to happen in an empty directory, such that we can assume `*` represents all out paths.

The issue is on the CLI `./.` is CWD, and cannot be changed, so there is no way to have an empty directory, and still having `./.` refer to somewhere else.

Which is why it now forcibly builds `<Tow-Boot/default.nix>`